File: killrvideo-service-videocatalog/src/test/java/com/killrvideo/service/video/grpc/VideoCatalogServiceGrpcValidatorTest.java
package com.killrvideo.service.video.grpc;

import io.grpc.stub.StreamObserver;
import killrvideo.common.CommonTypes;
import killrvideo.video_catalog.VideoCatalogServiceOuterClass.*;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static com.killrvideo.utils.GrpcMappingUtils.randomUuid;
import static org.mockito.Mockito.*;

class VideoCatalogServiceGrpcValidatorTest {
    private final VideoCatalogServiceGrpcValidator validator = new VideoCatalogServiceGrpcValidator();
    private StreamObserver<?> streamObserver;

    @BeforeEach
    public void setUp() {
        this.streamObserver = mock(StreamObserver.class);
    }

    @Test
    public void testValidateGrpcRequest_submitYoutubeVideo_Success() {
        SubmitYouTubeVideoRequest request = SubmitYouTubeVideoRequest.newBuilder()
                .setVideoId(randomUuid())
                .setUserId(randomUuid())
                .setName("Game")
                .setDescription("Game")
                .setYouTubeVideoId("xyz123")
                .build();
        validator.validateGrpcRequest_submitYoutubeVideo(request, streamObserver);

        verifySuccess();
    }

    @Test
    public void testValidateGrpcRequest_submitYoutubeVideo_Failure() {
        SubmitYouTubeVideoRequest request = SubmitYouTubeVideoRequest.getDefaultInstance();

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                validator.validateGrpcRequest_submitYoutubeVideo(request, streamObserver)
        );

        verifyFailure();
    }

    @Test
    public void testValidateGrpcRequest_getLatestPreviews_Success() {
        GetLatestVideoPreviewsRequest request = GetLatestVideoPreviewsRequest.newBuilder()
                .setPageSize(2)
                .build();
        validator.validateGrpcRequest_getLatestPreviews(request, streamObserver);

        verifySuccess();
    }

    @Test
    public void testValidateGrpcRequest_getLatestPreviews_Failure() {
        GetLatestVideoPreviewsRequest request = GetLatestVideoPreviewsRequest.getDefaultInstance();

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                validator.validateGrpcRequest_getLatestPreviews(request, streamObserver)
        );

        verifyFailure();
    }

    @Test
    public void testValidateGrpcRequest_getVideo_Success() {
        GetVideoRequest request = GetVideoRequest.newBuilder()
                .setVideoId(randomUuid())
                .build();
        validator.validateGrpcRequest_getVideo(request, streamObserver);

        verifySuccess();
    }

    @Test
    public void testValidateGrpcRequest_getVideo_Failure() {
        GetVideoRequest request = GetVideoRequest.getDefaultInstance();

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                validator.validateGrpcRequest_getVideo(request, streamObserver)
        );

        verifyFailure();
    }

    @Test
    public void testValidateGrpcRequest_getVideoPreviews_Success() {
        GetVideoPreviewsRequest request = GetVideoPreviewsRequest.newBuilder()
                .addVideoIds(randomUuid())
                .build();

        validator.validateGrpcRequest_getVideoPreviews(request, streamObserver);

        verifySuccess();
    }

    @Test
    public void testValidateGrpcRequest_getVideoPreviews_Failure() {
        List<CommonTypes.Uuid> videoIds = IntStream.range(0, 21)
                .mapToObj(i -> randomUuid()).collect(Collectors.toList());

        GetVideoPreviewsRequest request = GetVideoPreviewsRequest.newBuilder()
                .addAllVideoIds(videoIds)
                .build();

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                validator.validateGrpcRequest_getVideoPreviews(request, streamObserver)
        );

        verifyFailure();
    }

    @Test
    public void testValidateGrpcRequest_getUserVideoPreviews_Success() {
        GetUserVideoPreviewsRequest request = GetUserVideoPreviewsRequest.newBuilder()
                .setUserId(randomUuid())
                .setPageSize(2)
                .build();
        validator.validateGrpcRequest_getUserVideoPreviews(request, streamObserver);

        verifySuccess();
    }

    @Test
    public void testValidateGrpcRequest_getUserVideoPreviews_Failure() {
        GetUserVideoPreviewsRequest request = GetUserVideoPreviewsRequest.getDefaultInstance();

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                validator.validateGrpcRequest_getUserVideoPreviews(request, streamObserver)
        );

        verifyFailure();
    }

    private void verifySuccess() {
        verify(streamObserver, times(0)).onError(any());
        verify(streamObserver, times(0)).onCompleted();
    }

    private void verifyFailure() {
        verify(streamObserver, times(1)).onError(any());
        verify(streamObserver, times(1)).onCompleted();
    }
}

File: killrvideo-service-videocatalog/src/test/kotlin/com/killrvideo/service/video/grpc/VideoCatalogServiceGrpcValidatorTest.kt
package com.killrvideo.service.video.grpc

import com.killrvideo.utils.GrpcMappingUtils.randomUuid
import killrvideo.video_catalog.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.util.stream.IntStream

internal class VideoCatalogServiceGrpcValidatorTest {
    private val validator = VideoCatalogServiceGrpcValidator()

    @Test
    fun testValidateGrpcRequest_submitYoutubeVideo_Success() {
        val request = submitYouTubeVideoRequest {
            videoId = randomUuid()
            userId = randomUuid()
            name = "Game"
            description = "Game"
            youTubeVideoId = "xyz123"
        }
        validator.validateGrpcRequest_submitYoutubeVideo(request)
    }

    @Test
    fun testValidateGrpcRequest_submitYoutubeVideo_Failure() {
        val request = submitYouTubeVideoRequest {}
        assertThrows<IllegalArgumentException> {
            validator.validateGrpcRequest_submitYoutubeVideo(request)
        }
    }

    @Test
    fun testValidateGrpcRequest_getLatestPreviews_Success() {
        val request = getLatestVideoPreviewsRequest {
            pageSize = 2
        }
        validator.validateGrpcRequest_getLatestPreviews(request)
    }

    @Test
    fun testValidateGrpcRequest_getLatestPreviews_Failure() {
        val request = getLatestVideoPreviewsRequest {}
        assertThrows<IllegalArgumentException> {
            validator.validateGrpcRequest_getLatestPreviews(request)
        }
    }

    @Test
    fun testValidateGrpcRequest_getVideo_Success() {
        val request = getVideoRequest {
            videoId = randomUuid()
        }
        validator.validateGrpcRequest_getVideo(request)
    }

    @Test
    fun testValidateGrpcRequest_getVideo_Failure() {
        val request = getVideoRequest {}
        assertThrows<IllegalArgumentException> {
            validator.validateGrpcRequest_getVideo(request)
        }
    }

    @Test
    fun testValidateGrpcRequest_getVideoPreviews_Success() {
        val request = getVideoPreviewsRequest {
            videoIds.add(randomUuid())
        }
        validator.validateGrpcRequest_getVideoPreviews(request)
    }

    @Test
    fun testValidateGrpcRequest_getVideoPreviews_Failure() {
        val request = getVideoPreviewsRequest {
            IntStream.range(0, 21)
                .mapToObj { randomUuid() }
                .forEach {videoIds.add(it) }
        }
        assertThrows<IllegalArgumentException> {
            validator.validateGrpcRequest_getVideoPreviews(request)
        }
    }

    @Test
    fun testValidateGrpcRequest_getUserVideoPreviews_Success() {
        val request = getUserVideoPreviewsRequest {
            userId =randomUuid()
            pageSize = 2
        }
        validator.validateGrpcRequest_getUserVideoPreviews(request)

    }

    @Test
    fun testValidateGrpcRequest_getUserVideoPreviews_Failure() {
        val request = getUserVideoPreviewsRequest {}
        assertThrows<IllegalArgumentException> {
            validator.validateGrpcRequest_getUserVideoPreviews(request)
        }
    }
}
